AIP-84 | Add Auth for Import Error

jason810496 opened the pull request:
related: #42360 

### What

I have added the `requires_access_dag` dependency for the Import Error endpoints as well.

~However, the implementation still depends on the ongoing discussion in https://github.com/apache/airflow/pull/47062#discussion_r1975669524. ( I'm not sure if such granularity is needed for the Simple Auth Manager. If it is, I will add the necessary logic for `get_readable_dags` as well.  )~

~For reference, here are the Import Error implementations in the Legacy API:  
https://github.com/apache/airflow/blob/v2-10-test/airflow/api_connexion/endpoints/import_error_endpoint.py~

**Update**:
Yes, we should respect `readable_dag_ids`, so this PR is still depends on [AIP-84 | Add Auth for Dags #47433](https://github.com/apache/airflow/pull/47433) as it introduce `ReadableDagsFilterDep`.

jason810496 committed:
AIP-84 | Add Auth for Import Error

pierrejeambrun reviewed (commented):
Looks good, need the dagId filter indeed.

Lee-W committed:
fix: remove outdated requires_access_view

jason810496 committed:
Add permitted dags with import_error

jason810496 committed:
Add test for import error

Lee-W commented on airflow/api_fastapi/core_api/routes/public/import_error.py:
```diff
@@ -36,18 +42,38 @@
 
 
+def get_file_dag_ids(session: Session, filename: str) -> set[str]:
+    return {dag_id for dag_id in session.scalars(select(DagModel.dag_id).where(DagModel.fileloc == filename))}
```
Do we need to use `set` here? `dag_id` suppose to be unique

jason810496 replied:
We need `readable_dag_ids.intersection` and `file_dag_ids.issubset` some set operation in https://github.com/jason810496/airflow/blob/feature/AIP-84/auth/import_error/airflow/api_fastapi/core_api/routes/public/import_error.py#L91-L101

Lee-W replied:
then probably something like `set(session...)`

jason810496 replied:
Agree, thank for the suggestion.

Lee-W commented on airflow/api_fastapi/core_api/routes/public/import_error.py:
```diff
@@ -83,16 +131,48 @@ def get_import_errors(
 ) -> ImportErrorCollectionResponse:
     """Get all import errors."""
+    statement = select(ParseImportError)
     import_errors_select, total_entries = paginated_select(
```
this can be removed

Lee-W commented on airflow/api_fastapi/core_api/routes/public/import_error.py:
```diff
@@ -83,16 +131,48 @@ def get_import_errors(
+            offset=offset,
+            limit=limit,
+            session=session,
+        )
```
```suggestion
    import_errors_select, total_entries = paginated_select(
        statement=statement,
        order_by=order_by,
        offset=offset,
        limit=limit,
        session=session,
    )
```

pierrejeambrun replied:
+1

Lee-W commented on airflow/api_fastapi/core_api/routes/public/import_error.py:
```diff
@@ -83,16 +131,48 @@ def get_import_errors(
+    if not can_read_all_dags:
+        # if the user doesn't have access to all DAGs, only display errors from visible DAGs
+        readable_dag_ids = auth_manager.get_permitted_dag_ids(method="GET", user=user)
+        dagfiles_stmt = select(DagModel.fileloc).distinct().where(DagModel.dag_id.in_(readable_dag_ids))
```
Do you think we can also select DagModel.id here so that we don't need to select again in line 165?

jason810496 replied:
Sure, I just added a subquery to get corresponding `file_dag_ids` ( list of `dag_id` ) for each `DagModel .fileloc` in single query.

pierrejeambrun commented on airflow/api_fastapi/core_api/routes/public/import_error.py:
```diff
@@ -36,18 +42,38 @@
+    responses=create_openapi_http_exception_doc([status.HTTP_403_FORBIDDEN, status.HTTP_404_NOT_FOUND]),
+    dependencies=[
+        Depends(requires_access_view(AccessView.IMPORT_ERRORS)),
+        Depends(requires_access_dag("GET")),
```
I think we need to remove this 
```suggestion
```

Unlike other endpoints this is also check directly in the route implementation and handled differently:
- can_read_all_dags = auth_manager.is_authorized_dag(method="GET", user=user)

jason810496 replied:
Agree, I remove the `requires_access_dag` in both route and refactor the whole logic ( early return if user can access all DAGs )

pierrejeambrun commented on airflow/api_fastapi/core_api/routes/public/import_error.py:
```diff
@@ -36,18 +42,38 @@
 @import_error_router.get(
     "/{import_error_id}",
-    responses=create_openapi_http_exception_doc([status.HTTP_404_NOT_FOUND]),
+    responses=create_openapi_http_exception_doc([status.HTTP_403_FORBIDDEN, status.HTTP_404_NOT_FOUND]),
```
```suggestion
    responses=create_openapi_http_exception_doc([status.HTTP_404_NOT_FOUND]),
```

Forbidden is already part of the base router.

Lee-W commented on tests/api_fastapi/core_api/routes/public/test_import_error.py:
```diff
@@ -129,6 +140,53 @@ def test_get_import_error(
         assert response.json() == expected_json
 
+    def test_should_raises_401_unauthenticated(self, unauthenticated_test_client, setup):
+        import_error_id = setup[FILENAME1].id
```
It is not actually something introduced in this PR, but `setup` should probably be refactored into the fixture. it's quite unreadable now.

jason810496 replied:
I just refactored the whole test case, thanks to pointing out!.

Lee-W commented on tests/api_fastapi/core_api/routes/public/test_import_error.py:
```diff
@@ -48,6 +50,15 @@ class TestImportErrorEndpoint:
         clear_db_import_errors()
+        clear_db_dags()
+
+    @provide_session
```
I'm not sure whether we need it here. let see how discussion https://apache-airflow.slack.com/archives/C06K9Q5G2UA/p1741947155869659 goes

jason810496 replied:
Leave it as original since we doesn't restrict the `provide_session` decorator and `session` fixture in test usage.

Lee-W replied:
I actually like session fixture more, but we could keep it as it is till the next time we want to bring this topic up

pierrejeambrun commented on airflow/api_fastapi/core_api/routes/public/import_error.py:
```diff
@@ -56,12 +82,34 @@ def get_import_error(
     "",
+    dependencies=[
+        Depends(requires_access_view(AccessView.IMPORT_ERRORS)),
+        Depends(requires_access_dag("GET")),
```
```suggestion
```
Same, you can access the endpoint if you don't have access to all dags. That's checked manually inside the body of the route.

pierrejeambrun commented on tests/api_fastapi/core_api/routes/public/test_import_error.py:
```diff
@@ -129,6 +140,53 @@ def test_get_import_error(
+        mock_is_authorized_dag.assert_called_once_with(method="GET", user=mock.ANY)
+        mock_get_permitted_dag_ids.assert_called_once_with(user=mock.ANY)
+        assert response.status_code == 403
+        assert response.json() == {"detail": "You do not have read permission on any of the DAGs in the file"}
```
```suggestion
        # Arrange
        import_error_id = setup[FILENAME1].id
        mock_is_authorized_dag = mock_get_auth_manager.return_value.is_authorized_dag
        mock_is_authorized_dag.return_value = False
        mock_get_permitted_dag_ids = mock_get_auth_manager.return_value.get_permitted_dag_ids
        mock_get_permitted_dag_ids.return_value = set()
        
        # Act
        response = test_client.get(f"/public/importErrors/{import_error_id}")
        
        # Assert
        mock_is_authorized_dag.assert_called_once_with(method="GET", user=mock.ANY)
        mock_get_permitted_dag_ids.assert_called_once_with(user=mock.ANY)
        assert response.status_code == 403
        assert response.json() == {"detail": "You do not have read permission on any of the DAGs in the file"}
```

pierrejeambrun replied:
When tests are big with a lot of line to prepare / mock etc... You can structure them a little bit like this, or at least leave blank lines, it wil be easier to read.

That is a nitpick, non blocking.

pierrejeambrun reviewed (commented):
Nice, just a few suggestions, but looking good.

jason810496 committed:
Refactor import_error
- Remove requires_access_dag depends
- Refactor get_file_dag_ids helper
- Rename get_permitted_dag_ids to get_authorized_dag_ids
- Refactor get_import_errors
    - Early return if the user has access to all DAGs
    - Add subquery to get corresponding file_dag_ids for each fileloc in single query

jason810496 committed:
Refactor test_import_error
- Make test more readable by adding separate comment
- Add set auth_manager attribute helper
- Refactor test setup

jason810496 committed:
Remove get_file_dag_ids helper

jason810496 committed:
Remove 403 in api doc, early return for get_import_error router

Lee-W commented on tests/api_fastapi/core_api/routes/public/test_import_error.py:
```diff
@@ -48,46 +50,82 @@ class TestImportErrorEndpoint:
+    ) -> mock.Mock:
+        mock_batch_is_authorized_dag = mock_auth_manager.return_value.batch_is_authorized_dag
+        mock_batch_is_authorized_dag.return_value = batch_is_authorized_dag_return_value
+        return mock_batch_is_authorized_dag
```
we probably should make these static method fixtures

Lee-W commented on tests/api_fastapi/core_api/routes/public/test_import_error.py:
```diff
@@ -48,46 +50,82 @@ class TestImportErrorEndpoint:
+        return mock_batch_is_authorized_dag
+
+    @provide_session
+    def prepare_dag_model(self, session=None):
```
```suggestion
    def prepare_dag_model(self, session: Session=None) -> None:
```

Lee-W commented on tests/api_fastapi/core_api/routes/public/test_import_error.py:
```diff
@@ -48,46 +50,82 @@ class TestImportErrorEndpoint:
     @pytest.fixture(autouse=True)
     @provide_session
-    def setup(self, session=None) -> dict[str, ParseImportError]:
+    def setup(self, session=None):
```
```suggestion
    def setup(self, session: Session|None=None) -> None:
```

Lee-W reviewed (commented):
I did a quick review. The test is failing. I'll try to fix it later today if it hasn't been addressed yet. Thanks!

Lee-W committed:
test(api_fastapi): refactor test cases

Lee-W committed:
feat(import_error): rewrite groupby logic as array_agg is only supported in pg

Lee-W committed:
style: fix type checking
